Fix -Wunused-function suppression in Vates/VatesAPI/test/MockObjects.h

<!-- TEMPLATE FOR BUG REPORTS -->

### Expected behavior

Tried to suppress `-Wunused-function` warnings that appear in [Vates/VatesAPI/test/MockObjects.h](https://github.com/mantidproject/mantid/blob/3dd98feda21443932ecac1d9c437c02600104407/Vates/VatesAPI/test/MockObjects.h#L40) with GCC 5+

### Actual behavior

Despite the #pragmas being present, warnings are not suppressed (see Fedora 23 or Ubuntu 16.04 builds on Jenkins)

### Steps to reproduce the behavior

build VatesAPITest.

### Platforms affected

LInux w/ GCC 5 or later (Fedora 23+, Ubuntu 16.04+, Arch)

Is this an issue with GCC?

Compiling the following short program on Ubuntu 16.04 with GCC 5.3 and Clang 3.8

```
#include <string>

#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wunused-function"

static std::string getString() {
  return "hello world!";
}

#pragma GCC diagnostic pop

int main()
{}
```
```
$ clang++ test.cpp -Wall
$ 
```
```
$ g++ test.cpp -Wall
test.cpp:6:20: warning: ‘std::__cxx11::string getString()’ defined but not used [-Wunused-function]
 static std::string getString() {
                    ^
$ 
```

Without the #pragmas

```
$ clang++ test.cpp -Wall
test.cpp:6:20: warning: unused function 'getString' [-Wunused-function]
static std::string getString() {
                   ^
1 warning generated.
$
```

```
$ g++ test.cpp -Wall
test.cpp:6:20: warning: ‘std::__cxx11::string getString()’ defined but not used [-Wunused-function]
 static std::string getString() {
                    ^
$ 
```

--- a/Framework/Kernel/inc/MantidKernel/WarningSuppressions.h
+++ b/Framework/Kernel/inc/MantidKernel/WarningSuppressions.h
@@ -89,4 +89,10 @@
 #endif
 // clang-format on
 
+#ifdef GCC_VERSION
+#define GCC_UNUSED_FUNCTION __attribute__((unused))
+#else
+#define GCC_UNUSED_FUNCTION
+#endif
+
 #endif /*MANTID_KERNEL_WARNINGSUPPRESSIONS_H_*/

--- a/Vates/VatesAPI/test/MockObjects.h
+++ b/Vates/VatesAPI/test/MockObjects.h
@@ -31,17 +31,6 @@ using Mantid::Geometry::MDHistoDimension;
 using Mantid::Geometry::MDHistoDimension_sptr;
 using Mantid::coord_t;
 
-// Allow unused functions.
-#if __clang__
-#pragma clang diagnostic push
-#pragma clang diagnostic ignored "-Wunused-function"
-#endif
-
-#if defined(GCC_VERSION) && GCC_VERSION >= 50000
-#pragma GCC diagnostic push
-#pragma GCC diagnostic ignored "-Wunused-function"
-#endif
-
 //=====================================================================================
 // Test Helper Types. These are shared by several tests in VatesAPI
 //=====================================================================================
@@ -195,7 +184,8 @@ Create a field data entry containing (as contents) the argument text.
 @param testData : Text to enter
 @return new vtkFieldData object containing text.
 */
-vtkFieldData *createFieldDataWithCharArray(std::string testData) {
+GCC_UNUSED_FUNCTION vtkFieldData *
+createFieldDataWithCharArray(std::string testData) {
   vtkFieldData *fieldData = vtkFieldData::New();
   vtkCharArray *charArray = vtkCharArray::New();
   charArray->SetName(Mantid::VATES::XMLDefinitions::metaDataId().c_str());
@@ -279,10 +269,11 @@ view.
 view.
 @return full xml as string.
 */
-std::string constructXML(const std::string &xDimensionIdMapping,
-                         const std::string &yDimensionIdMapping,
-                         const std::string &zDimensionIdMapping,
-                         const std::string &tDimensionIdMapping) {
+GCC_UNUSED_FUNCTION std::string
+constructXML(const std::string &xDimensionIdMapping,
+             const std::string &yDimensionIdMapping,
+             const std::string &zDimensionIdMapping,
+             const std::string &tDimensionIdMapping) {
   return std::string("<?xml version=\"1.0\" encoding=\"utf-8\"?>") +
          "<MDInstruction>" + "<MDWorkspaceName>Input</MDWorkspaceName>" +
          "<MDWorkspaceLocation>test_horace_reader.sqw</MDWorkspaceLocation>" +
@@ -376,7 +367,7 @@ view.
 view.
 @return full xml as string.
 */
-std::string
+GCC_UNUSED_FUNCTION std::string
 constructXMLForMDEvHelperData(const std::string &xDimensionIdMapping,
                               const std::string &yDimensionIdMapping,
                               const std::string &zDimensionIdMapping,
@@ -407,8 +398,8 @@ Mantid::API::Workspace_sptr createSimple3DWorkspace() {
   return outWs;
 }
 
-Mantid::API::Workspace_sptr get3DWorkspace(bool integratedTDimension,
-                                           bool sliceMD) {
+GCC_UNUSED_FUNCTION Mantid::API::Workspace_sptr
+get3DWorkspace(bool integratedTDimension, bool sliceMD) {
   using namespace Mantid::API;
   using namespace Mantid::DataObjects;
 
@@ -446,7 +437,8 @@ Mantid::API::Workspace_sptr get3DWorkspace(bool integratedTDimension,
  * @param fieldName : The requested field data entry
  * @return The value of the requested field data entry
  */
-std::string getStringFieldDataValue(vtkDataSet *ds, std::string fieldName) {
+GCC_UNUSED_FUNCTION std::string getStringFieldDataValue(vtkDataSet *ds,
+                                                        std::string fieldName) {
   vtkAbstractArray *value =
       ds->GetFieldData()->GetAbstractArray(fieldName.c_str());
   vtkStringArray *array = vtkStringArray::SafeDownCast(value);
@@ -454,13 +446,5 @@ std::string getStringFieldDataValue(vtkDataSet *ds, std::string fieldName) {
 }
 
 } // namespace
-
-#if __clang__
-#pragma clang diagnostic pop
-#endif
-
-#if defined(GCC_VERSION) && GCC_VERSION >= 50000
-#pragma GCC diagnostic pop
-#endif
 
 #endif

--- a/Vates/VatesAPI/test/MockObjects.h
+++ b/Vates/VatesAPI/test/MockObjects.h
@@ -379,7 +379,6 @@ std::string constructXMLForMDEvHelperData(
              xDimensionIdMapping, yDimensionIdMapping, zDimensionIdMapping,
              tDimensionIdMapping) +
          "</MDInstruction>";
-  `
 }
 
 Mantid::API::Workspace_sptr createSimple3DWorkspace() {

--- a/Vates/VatesAPI/test/MockObjects.h
+++ b/Vates/VatesAPI/test/MockObjects.h
@@ -379,6 +379,7 @@ std::string constructXMLForMDEvHelperData(
              xDimensionIdMapping, yDimensionIdMapping, zDimensionIdMapping,
              tDimensionIdMapping) +
          "</MDInstruction>";
+  `
 }
 
 Mantid::API::Workspace_sptr createSimple3DWorkspace() {

--- a/Vates/VatesAPI/test/MockObjects.h
+++ b/Vates/VatesAPI/test/MockObjects.h
@@ -184,8 +184,8 @@ Create a field data entry containing (as contents) the argument text.
 @param testData : Text to enter
 @return new vtkFieldData object containing text.
 */
-vtkFieldData *
-createFieldDataWithCharArray(std::string testData) GCC_UNUSED_FUNCTION {
+GCC_UNUSED_FUNCTION vtkFieldData *
+createFieldDataWithCharArray(std::string testData) {
   vtkFieldData *fieldData = vtkFieldData::New();
   vtkCharArray *charArray = vtkCharArray::New();
   charArray->SetName(Mantid::VATES::XMLDefinitions::metaDataId().c_str());
@@ -269,11 +269,11 @@ view.
 view.
 @return full xml as string.
 */
-std::string
+GCC_UNUSED_FUNCTION std::string
 constructXML(const std::string &xDimensionIdMapping,
              const std::string &yDimensionIdMapping,
              const std::string &zDimensionIdMapping,
-             const std::string &tDimensionIdMapping) GCC_UNUSED_FUNCTION {
+             const std::string &tDimensionIdMapping) {
   return std::string("<?xml version=\"1.0\" encoding=\"utf-8\"?>") +
          "<MDInstruction>" + "<MDWorkspaceName>Input</MDWorkspaceName>" +
          "<MDWorkspaceLocation>test_horace_reader.sqw</MDWorkspaceLocation>" +
@@ -367,11 +367,11 @@ view.
 view.
 @return full xml as string.
 */
-std::string constructXMLForMDEvHelperData(
-    const std::string &xDimensionIdMapping,
-    const std::string &yDimensionIdMapping,
-    const std::string &zDimensionIdMapping,
-    const std::string &tDimensionIdMapping) GCC_UNUSED_FUNCTION {
+GCC_UNUSED_FUNCTION std::string
+constructXMLForMDEvHelperData(const std::string &xDimensionIdMapping,
+                              const std::string &yDimensionIdMapping,
+                              const std::string &zDimensionIdMapping,
+                              const std::string &tDimensionIdMapping) {
   return std::string("<?xml version=\"1.0\" encoding=\"utf-8\"?>") +
          "<MDInstruction>" + "<MDWorkspaceName>Input</MDWorkspaceName>" +
          "<MDWorkspaceLocation>test_horace_reader.sqw</MDWorkspaceLocation>" +
@@ -379,7 +379,6 @@ std::string constructXMLForMDEvHelperData(
              xDimensionIdMapping, yDimensionIdMapping, zDimensionIdMapping,
              tDimensionIdMapping) +
          "</MDInstruction>";
-  `
 }
 
 Mantid::API::Workspace_sptr createSimple3DWorkspace() {
@@ -399,8 +398,8 @@ Mantid::API::Workspace_sptr createSimple3DWorkspace() {
   return outWs;
 }
 
-Mantid::API::Workspace_sptr get3DWorkspace(bool integratedTDimension,
-                                           bool sliceMD) GCC_UNUSED_FUNCTION {
+GCC_UNUSED_FUNCTION Mantid::API::Workspace_sptr
+get3DWorkspace(bool integratedTDimension, bool sliceMD) {
   using namespace Mantid::API;
   using namespace Mantid::DataObjects;
 
@@ -438,8 +437,8 @@ Mantid::API::Workspace_sptr get3DWorkspace(bool integratedTDimension,
  * @param fieldName : The requested field data entry
  * @return The value of the requested field data entry
  */
-std::string getStringFieldDataValue(vtkDataSet *ds,
-                                    std::string fieldName) GCC_UNUSED_FUNCTION {
+GCC_UNUSED_FUNCTION std::string getStringFieldDataValue(vtkDataSet *ds,
+                                                        std::string fieldName) {
   vtkAbstractArray *value =
       ds->GetFieldData()->GetAbstractArray(fieldName.c_str());
   vtkStringArray *array = vtkStringArray::SafeDownCast(value);

--- a/Vates/VatesAPI/test/MockObjects.h
+++ b/Vates/VatesAPI/test/MockObjects.h
@@ -184,8 +184,8 @@ Create a field data entry containing (as contents) the argument text.
 @param testData : Text to enter
 @return new vtkFieldData object containing text.
 */
-GCC_UNUSED_FUNCTION vtkFieldData *
-createFieldDataWithCharArray(std::string testData) {
+vtkFieldData *
+createFieldDataWithCharArray(std::string testData) GCC_UNUSED_FUNCTION {
   vtkFieldData *fieldData = vtkFieldData::New();
   vtkCharArray *charArray = vtkCharArray::New();
   charArray->SetName(Mantid::VATES::XMLDefinitions::metaDataId().c_str());
@@ -269,11 +269,11 @@ view.
 view.
 @return full xml as string.
 */
-GCC_UNUSED_FUNCTION std::string
+std::string
 constructXML(const std::string &xDimensionIdMapping,
              const std::string &yDimensionIdMapping,
              const std::string &zDimensionIdMapping,
-             const std::string &tDimensionIdMapping) {
+             const std::string &tDimensionIdMapping) GCC_UNUSED_FUNCTION {
   return std::string("<?xml version=\"1.0\" encoding=\"utf-8\"?>") +
          "<MDInstruction>" + "<MDWorkspaceName>Input</MDWorkspaceName>" +
          "<MDWorkspaceLocation>test_horace_reader.sqw</MDWorkspaceLocation>" +
@@ -367,11 +367,11 @@ view.
 view.
 @return full xml as string.
 */
-GCC_UNUSED_FUNCTION std::string
-constructXMLForMDEvHelperData(const std::string &xDimensionIdMapping,
-                              const std::string &yDimensionIdMapping,
-                              const std::string &zDimensionIdMapping,
-                              const std::string &tDimensionIdMapping) {
+std::string constructXMLForMDEvHelperData(
+    const std::string &xDimensionIdMapping,
+    const std::string &yDimensionIdMapping,
+    const std::string &zDimensionIdMapping,
+    const std::string &tDimensionIdMapping) GCC_UNUSED_FUNCTION {
   return std::string("<?xml version=\"1.0\" encoding=\"utf-8\"?>") +
          "<MDInstruction>" + "<MDWorkspaceName>Input</MDWorkspaceName>" +
          "<MDWorkspaceLocation>test_horace_reader.sqw</MDWorkspaceLocation>" +
@@ -379,6 +379,7 @@ constructXMLForMDEvHelperData(const std::string &xDimensionIdMapping,
              xDimensionIdMapping, yDimensionIdMapping, zDimensionIdMapping,
              tDimensionIdMapping) +
          "</MDInstruction>";
+  `
 }
 
 Mantid::API::Workspace_sptr createSimple3DWorkspace() {
@@ -398,8 +399,8 @@ Mantid::API::Workspace_sptr createSimple3DWorkspace() {
   return outWs;
 }
 
-GCC_UNUSED_FUNCTION Mantid::API::Workspace_sptr
-get3DWorkspace(bool integratedTDimension, bool sliceMD) {
+Mantid::API::Workspace_sptr get3DWorkspace(bool integratedTDimension,
+                                           bool sliceMD) GCC_UNUSED_FUNCTION {
   using namespace Mantid::API;
   using namespace Mantid::DataObjects;
 
@@ -437,8 +438,8 @@ get3DWorkspace(bool integratedTDimension, bool sliceMD) {
  * @param fieldName : The requested field data entry
  * @return The value of the requested field data entry
  */
-GCC_UNUSED_FUNCTION std::string getStringFieldDataValue(vtkDataSet *ds,
-                                                        std::string fieldName) {
+std::string getStringFieldDataValue(vtkDataSet *ds,
+                                    std::string fieldName) GCC_UNUSED_FUNCTION {
   vtkAbstractArray *value =
       ds->GetFieldData()->GetAbstractArray(fieldName.c_str());
   vtkStringArray *array = vtkStringArray::SafeDownCast(value);

